plugins/hls-change-type-signature-plugin/src/Ide/Plugin/ChangeTypeSignature.hs

```diff
@@ -0,0 +1,247 @@
+                         -- | the location of the declaration signature
+                         declSrcSpan  :: RealSrcSpan
+                                       ,
+                         -- | the diagnostic to solve
+                         diagnostic   :: Diagnostic
+                                       }
+
+-- | Constraint needed to trackdown OccNames in signatures
+type SigName = (HasOccName (IdP GhcPs))
+
+
+-- | Create a CodeAction from a Diagnostic
+generateAction :: SigName => Uri -> [LHsDecl GhcPs] -> Diagnostic -> Maybe (Command |? CodeAction)
+generateAction uri decls diag = changeSigToCodeAction uri <$> diagnosticToChangeSig decls diag
+
+-- | Convert a diagnostic into a ChangeSignature and add the proper SrcSpan
+diagnosticToChangeSig :: SigName => [LHsDecl GhcPs] -> Diagnostic -> Maybe ChangeSignature
+diagnosticToChangeSig decls diagnostic = do
+    -- regex match on the GHC Error Message
+    (expectedType, actualType', declName) <- matchingDiagnostic diagnostic
+    let actualType = tidyTypeSignature actualType'
+    -- Find the definition and it's location
+    (declSrcSpan, ghcSig) <- findSigLocOfStringDecl decls (T.unpack declName)
+    -- Make sure the given "Actual Type" is a full signature
+    isValidMessage expectedType ghcSig
+    pure $ ChangeSignature{..}
+
+-- | Does the GHC Error Message give us a Signature we can use?
+-- We only want to change signatures when the "expected signature" given by GHC
+-- matches the one given by the user
+isValidMessage :: DefinedSig -> ExpectedSig -> Maybe ()
+isValidMessage defSig expSig = if defSig == expSig then Just () else Nothing
+
+-- | If a diagnostic has the proper message create a ChangeSignature from it
+matchingDiagnostic :: Diagnostic -> Maybe (ExpectedSig, ActualSig, DeclName)
+matchingDiagnostic diag@Diagnostic{_message} = case map (unwrapMatch . (=~) _message) errorMessageRegexes of
+                                                       []  -> Nothing
+                                                       css -> join $ find isJust css
+    where
+        unwrapMatch :: (Text, Text, Text, [Text]) -> Maybe (ExpectedSig, ActualSig, DeclName)
+        -- due to using (.|\n) in regex we have to drop the erroneous, but necessary ("." doesn't match newlines), match
+        unwrapMatch (_, _, _, [exp, act, _, name]) = Just (exp, act, name)
+        unwrapMatch _                              = Nothing
+
+-- | List of regexes that match various Error Messages
+errorMessageRegexes :: [Text]
+errorMessageRegexes = [ -- be sure to add new Error Messages Regexes at the bottom to not fail any existing tests
+    "Expected type: (" <> typeSigRegex <> ")\n +Actual type: (" <> typeSigRegex <> ")\n(.|\n)+In an equation for ‘(.+)’"
+    , "Couldn't match expected type ‘(" <> typeSigRegex <> ")’ with actual type ‘(" <> typeSigRegex <> ")’\n(.|\n)+In an equation for ‘(.+)’"
+    ]
+    where
+        -- The brackets in position 1 and 2 MUST BE IN THAT ORDER
```

Comment:
Mysterious.

Reply:
Yeah, I reached out to the maintainers of regex-tdfa. I was always under the assumption that you could escape things in bracket expressions, but apparently not.